libcst/tests/test_exceptions.py

```diff
@@ -4,54 +4,90 @@
 # LICENSE file in the root directory of this source tree.
 
 # pyre-strict
+
+import pickle
 from textwrap import dedent
 
-from libcst._exceptions import ParserSyntaxError
+import libcst as cst
 from libcst.testing.utils import UnitTest, data_provider
 
 
 class ExceptionsTest(UnitTest):
     @data_provider(
-        [
-            (
-                ParserSyntaxError(
-                    message="some message",
-                    encountered=None,  # EOF
-                    expected=None,  # EOF
-                    pos=(1, 0),
-                    lines=["abcd"],
+        {
+            "simple": (
+                cst.ParserSyntaxError(
+                    "some message", lines=["abcd"], raw_line=1, raw_column=0
                 ),
                 dedent(
                     """
-                    Syntax Error: some message @ 1:1.
-                    Encountered end of file (EOF), but expected end of file (EOF).
+                    Syntax Error @ 1:1.
+                    some message
 
                     abcd
                     ^
                     """
                 ).strip(),
             ),
-            (
-                ParserSyntaxError(
-                    message="some message",
-                    encountered="encountered_value",
-                    expected=["expected_value"],
-                    pos=(1, 2),
-                    lines=["\tabcd\r\n"],
+            "tab_expansion": (
+                cst.ParserSyntaxError(
+                    "some message", lines=["\tabcd\r\n"], raw_line=1, raw_column=2
                 ),
                 dedent(
                     """
-                    Syntax Error: some message @ 1:10.
-                    Encountered 'encountered_value', but expected one of ['expected_value'].
+                    Syntax Error @ 1:10.
+                    some message
 
                             abcd
                              ^
                     """
                 ).strip(),
             ),
-        ]
+            "shows_last_line_with_text": (
+                cst.ParserSyntaxError(
+                    "some message",
+                    lines=["abcd\n", "efgh\n", "\n", "\n", "\n", "\n", "\n"],
+                    raw_line=5,
+                    raw_column=0,
+                ),
+                dedent(
+                    """
+                    Syntax Error @ 5:1.
+                    some message
+
+                    efgh
+                        ^
+                    """
+                ).strip(),
+            ),
+            "empty_file": (
+                cst.ParserSyntaxError(
+                    "some message", lines=[""], raw_line=1, raw_column=0
+                ),
+                dedent(
+                    """
+                    Syntax Error @ 1:1.
+                    some message
```

Comment:
Why this case didn't show the position indicator `^`?

Reply:
I added some logic that suppresses the context if it would just show an empty line, because I found an empty line followed by a caret indicator really confusing.

If you think this behavior is more confusing, I can remove it and unconditionally show the caret.

I'll add a comment to the test case to clarify the behavior a bit more.